### ISSUE TITLE
build cockroachdb for mips64el platform

### ISSUE THREAD

**Issue by vans88** [cockroachdb/cockroach#10778]
Hello everyone, I wanna compile `cockroachdb` from source, but there is a question that makes me feel puzzled. Firstly, I will show you some information about my platform.

## Platform info
```
[root@dscn2 ~]# uname -a
Linux dscn2 4.4.21+ #3 SMP PREEMPT Wed Oct 12 09:14:23 CST 2016 mips64 mips64 mips64 GNU/Linux
[root@dscn2 ~]# lscpu
Architecture:          mips64
Byte Order:            Little Endian
CPU(s):                15
On-line CPU(s) list:   0-14
Thread(s) per core:    1
Core(s) per socket:    7
Socket(s):             2
NUMA node(s):          4
.. ... ... .. ...
[root@dscn2 ~]# g++ -v
Using built-in specs.
COLLECT_GCC=/bin/g++
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/mips64el-redhat-linux/4.9.3/lto-wrapper
Target: mips64el-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-arch=loongson3a --enable-languages=c,c++,objc,obj-c++,fortran,go,lto --enable-plugin --disable-libgcj --with-isl=/root/rpmbuild/BUILD/gcc-4.9.3/obj-mips64el-redhat-linux/isl-install --with-cloog=/root/rpmbuild/BUILD/gcc-4.9.3/obj-mips64el-redhat-linux/cloog-install --enable-gnu-indirect-function --with-long-double-128 --build=mips64el-redhat-linux
Thread model: posix
gcc version 4.9.3 20150626 (Red Hat 4.9.3-3) (GCC)
```
## OS info
```
[root@dscn2 ~]# cat /etc/issue
Fedora release 21 (Twenty One)
Kernel \r on an \m (\l)
```
## Error info
```
[root@dscn2 cockroach]# make install 
GOPATH set to /root/mygo
/root/mygo/bin added to PATH
go install -v
github.com/cockroachdb/cockroach/pkg/cli/cliflags
github.com/petermattis/goid
github.com/cockroachdb/cockroach/pkg/util/shuffle
github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror
github.com/cockroachdb/cockroach/pkg/util/bufalloc
github.com/cockroachdb/cockroach/pkg/util/syncutil
github.com/cockroachdb/cockroach/pkg/build
github.com/cockroachdb/c-jemalloc
github.com/cockroachdb/cockroach/pkg/util/log/logflags
github.com/cockroachdb/cockroach/pkg/util/humanizeutil
github.com/cockroachdb/cockroach/pkg/util/uuid
github.com/cockroachdb/cockroach/pkg/util/duration
github.com/cockroachdb/cockroach/pkg/util/decimal
github.com/cockroachdb/cockroach/pkg/util/httputil
# github.com/petermattis/goid
../../petermattis/goid/goid_go1.6plus.go:46: missing function body for "getg"
github.com/cockroachdb/cockroach/pkg/util/retry
github.com/cockroachdb/cmux
github.com/cockroachdb/cockroach/pkg/util/interval
github.com/cockroachdb/cockroach/pkg/sql/privilege
github.com/cockroachdb/c-protobuf
github.com/cockroachdb/c-snappy
github.com/cockroachdb/cockroach/pkg/ui
github.com/cockroachdb/cockroach/pkg/util/sdnotify
github.com/cockroachdb/pq
github.com/cockroachdb/cockroach/pkg/util/caller
github.com/cockroachdb/cockroach/pkg/util/protoutil
github.com/cockroachdb/cockroach/pkg/util/encoding
github.com/cockroachdb/c-rocksdb
# github.com/cockroachdb/c-rocksdb
### g++: error: unrecognized command line option ‘-momit-leaf-frame-pointer’
Makefile:79: recipe for target 'install' failed
make: *** [install] Error 2
```

**Comment by vans88** [comment]
By searching with the keyword "momit-leaf-frame-pointer", I obtain the following information:

## Related information
```
/root/mygo/src/github.com/cockroachdb
[root@dscn2 cockroachdb]# grep "momit-leaf-frame-pointer" . -rn
./c-rocksdb/internal/Makefile:89:# Skip for archs that don't support -momit-leaf-frame-pointer
./c-rocksdb/internal/Makefile:90:ifeq (,$(shell $(CXX) -fsyntax-only -momit-leaf-frame-pointer -xc /dev/null 2>&1))
./c-rocksdb/internal/Makefile:91:OPT += -momit-leaf-frame-pointer
./c-rocksdb/cgo_flags.go:14:// #cgo CXXFLAGS: -std=c++11 -fno-omit-frame-pointer -momit-leaf-frame-pointer
```

**Comment by vans88** [comment]
I modified the following line in file `c-rocksdb/cgo_flags.go:14`
### the original line
`#cgo CXXFLAGS: -std=c++11 -fno-omit-frame-pointer -momit-leaf-frame-pointer`
### after modified 
`#cgo CXXFLAGS: -std=c++11 -fno-omit-frame-pointer`

**Comment by vans88** [comment]
By Searching the source code, I find the following useful information:

### Error info
note:   no known conversion for argument 2 from ‘rocksdb::ImmutableCFOptions’ to ‘const rocksdb::Options&’


### call method
```
struct DBSstFileWriter {
  std::unique_ptr<rocksdb::Options> options;
  rocksdb::ImmutableCFOptions ioptions;
  rocksdb::SstFileWriter rep;

  DBSstFileWriter(rocksdb::Options* o)
      : options(o),
        ioptions(*o),
        rep(rocksdb::EnvOptions(), ioptions, o->comparator) {
  }
  virtual ~DBSstFileWriter() { }
};
```
### c-rocksdb/internal/include/rocksdb/sst_file_writer.h line num 50-55
```
class SstFileWriter {
 public:
  SstFileWriter(const EnvOptions& env_options, const Options& options,
                const Comparator* user_comparator);
  ~SstFileWriter();
... ... ... ... .. .. .. ... ... ... ... ... 
```

**Comment by petermattis** [comment]
@vans88 I believe you're seeing the error that is fixed in https://github.com/cockroachdb/cockroach/pull/9616. Note that CockroachDB is still using RocksDB 4.8.

**Comment by petermattis** [comment]
@vans88 We've upgraded to RocksDB 4.11.2 which necessitated fixing the compilation errors you were running into.

**Comment by petermattis** [comment]
I believe this was fixed with the RocksDB 4.11.2 upgrade. Please ping the issue again (or open a new one) if you're still seeing problems.
